R/as_model_out_df.R

```diff
@@ -0,0 +1,211 @@
+  }
+  if (!is.null(output_type_id_col)) {
+    output_type_id_col <- validate_col_input(output_type_id_col, call = call)
+    tbl <- rename_col(tbl, output_type_id_col, old_names, call)
+  }
+  if (!is.null(value_col)) {
+    value_col <- validate_col_input(value_col, call = call)
+    tbl <- rename_col(tbl, value_col, old_names, call)
+  }
+  return(tbl)
+}
+
+validate_col_input <- function(x, call = rlang::caller_env()) {
+  if (!is.character(x)) {
+    cli::cli_abort("{.arg {rlang::caller_arg(x)}} must be a {.cls character} string
+                       instead of {.cls {typeof(x)}}",
+      call = call
+    )
+  }
+  if (length(x) > 1L) {
+    cli::cli_warn("{.arg {rlang::caller_arg(x)}} must be character vector of
+                       length {.val {1L}} not length {.val {length(x)}}.
+                      Only element 1 ({.val {x[1]}}) used, the rest ignored.",
+      call = call
+    )
+    x <- x[1]
+  }
+  return(x)
+}
+
+rename_col <- function(x, col_name, old_names, call) {
+  arg_name <- rlang::caller_arg(col_name)
+  if (!col_name %in% names(x)) {
+    cli::cli_abort(c(
+      "x" = "{.arg {arg_name}} value {.val {col_name}}
+                       not a valid column name in {.arg x}",
+      "!" = "Must be one of {.val {old_names}}"
+    ),
+    call = call
+    )
+  } else {
+    new_col_name <- gsub("_col", "", arg_name)
+    names(x)[names(x) == col_name] <- new_col_name
+  }
+  return(x)
+}
+
+trim_tbl_to_task_ids <- function(tbl, task_id_cols, hub_con,
+                                 call = rlang::caller_env()) {
+  if (is.null(task_id_cols) && is.null(hub_con)) {
+    cli::cli_abort(
+      c(
+        "x" = "Cannot trim to task IDs when both
+                  {.arg task_id_cols} and {.arg hub_con} are NULL",
+        "!" = "Please supply appropriate input to either
+                {.arg task_id_cols} or {.arg hub_con}."
+      ),
+      call = call
+    )
+  } else {
+    if (!is.null(task_id_cols)) {
+      if (!all(task_id_cols %in% names(tbl))) {
+        cli::cli_alert_warning(
+          "{.arg task_id_cols} value{?s} {.val
+                  {task_id_cols[!task_id_cols %in% names(tbl)]}
+                  } not a valid {.arg tbl} column name. Ignored."
+        )
+      }
+      task_id_cols <- task_id_cols[task_id_cols %in% names(tbl)]
+    } else {
+      if (!inherits(hub_con, "hub_connection")) {
+        cli::cli_abort("{.arg hub_con} must be a valid object of
+                                   class {.cls hub_connection}",
+          call = call
+        )
+      }
+      task_id_cols <- get_task_id_names(attr(hub_con, "config_tasks"))
+    }
+  }
+
+  return(tbl[, c(task_id_cols, std_colnames)])
+}
+
+remove_model_out_tbl_empty <- function(tbl) {
+  tbl[, purrr::map_lgl(tbl, ~ !all(is.na(.x)))]
+}
+
+std_col_order_model_out_tbl <- function(tbl) {
+  tbl[
+    ,
+    c(
+      std_colnames[1],
+      names(tbl)[!names(tbl) %in% std_colnames],
+      std_colnames[-1]
+    )
+  ]
```

Comment:
This is speculative premature optimization -- but seemingly minor operations on giant data frames have led to noticeable slowdowns in the past, so will raise it here anyways.  should we only re-order columns if necessary?  something like this?
```suggestion
  out_cols <- c(
    std_colnames[1],
    names(tbl)[!names(tbl) %in% std_colnames],
    std_colnames[-1])
  if (isTRUE(all.equal(std_colnames, names(tbl)))) return(tbl)
  return(tbl[, out_cols])
```

Reply:
While indeed, theoretically, we could shave a bit of time with this approach in the case where tbl is already in the correct order, because base reordering of data.frames is generally very efficient already, it would likely only be in the order of a few microseconds or less. So firmly in the realm of a nano-optimisation.

In addition, the benchmark below shows that the added checking overhead appears to double the time required in the case where reordering is required (which I believe will be the majority of the cases).

Note as well that the time taken does not change when we double the size of the data.frame by boubling the number of columns:

Here's a quick benchmark:
``` r
test_df <- tibble::tibble(a = rnorm(10e7),
                          b = rnorm(10e7),
                          c = rnorm(10e7))

lobstr::obj_size(test_df)
#> 2.40 GB

order <- names(test_df)
rev_order <- rev(names(test_df))
random_order <- names(test_df)[sample.int(ncol(test_df), ncol(test_df))]

microbenchmark::microbenchmark(
    correct_order =
        {test_df[, order]},
    reverse_order = {
        test_df[, rev_order]},
    random_order = {
        test_df[, random_order]
    },
    check_first_correct_order = {
        if (isTRUE(all.equal(order, names(test_df)))) {} else {test_df[, order]}
    },
    check_first_reverse_order = {
        if (isTRUE(all.equal(rev_order, names(test_df)))) {} else {test_df[, rev_order]}
    },
    check_first_random_order = {
        if (isTRUE(all.equal(random_order, names(test_df)))) {} else {test_df[, random_order]}
    },
    ctrl = {
        test_df[, ]
    }
)
#> Warning in microbenchmark::microbenchmark(correct_order = {: less accurate
#> nanosecond times to avoid potential integer overflows
#> Unit: microseconds
#>                       expr    min     lq     mean  median      uq     max neval
#>              correct_order  7.421  7.585  7.91013  7.7080  7.9335  20.500   100
#>              reverse_order  7.339  7.544 10.11101  7.7490  7.9335 238.333   100
#>               random_order  7.339  7.503  7.77155  7.6465  7.8720  12.136   100
#>  check_first_correct_order  5.412  5.781  5.91876  5.8630  6.0270   7.790   100
#>  check_first_reverse_order 15.088 15.457 17.97973 15.6210 15.9080 230.297   100
#>   check_first_random_order 15.293 15.539 16.29627 15.7030 15.9490  56.375   100
#>                       ctrl  2.378  2.501  2.58136  2.5420  2.6240   3.608   100
#>  cld
#>   b 
#>   b 
#>   b 
#>  ab 
#>    c
#>    c
#>  a

test_df <- tibble::tibble(a = rnorm(10e7),
                          b = rnorm(10e7),
                          c = rnorm(10e7),
                          d = rnorm(10e7),
                          e = rnorm(10e7),
                          f = rnorm(10e7))

lobstr::obj_size(test_df)
#> 4.80 GB

order <- names(test_df)
rev_order <- rev(names(test_df))
random_order <- names(test_df)[sample.int(ncol(test_df), ncol(test_df))]

microbenchmark::microbenchmark(
    correct_order =
        {test_df[, order]},
    reverse_order = {
        test_df[, rev_order]},
    random_order = {
        test_df[, random_order]
    },
    check_first_correct_order = {
        if (isTRUE(all.equal(order, names(test_df)))) {} else {test_df[, order]}
    },
    check_first_reverse_order = {
        if (isTRUE(all.equal(rev_order, names(test_df)))) {} else {test_df[, rev_order]}
    },
    check_first_random_order = {
        if (isTRUE(all.equal(random_order, names(test_df)))) {} else {test_df[, random_order]}
    },
    ctrl = {
        test_df[, ]
    }
)
#> Unit: microseconds
#>                       expr    min     lq     mean  median      uq    max neval
#>              correct_order  7.544  7.749  7.98393  7.9130  8.0975 12.177   100
#>              reverse_order  7.380  7.626  8.20451  7.7900  8.0360 44.362   100
#>               random_order  7.339  7.667  7.85396  7.7900  7.9540  8.979   100
#>  check_first_correct_order  5.535  5.822  6.11802  5.9245  6.0680 20.664   100
#>  check_first_reverse_order 15.252 15.580 16.19418 15.7440 16.0105 30.340   100
#>   check_first_random_order 15.293 15.621 15.91497 15.7850 16.1130 19.680   100
#>                       ctrl  2.296  2.542  2.66787  2.6240  2.7060  5.904   100
#>   cld
#>    c 
#>    c 
#>    c 
#>   b  
#>     d
#>     d
#>  a
```

<sup>Created on 2023-06-28 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Worth noting that all this is happening in the realm of microseconds so not noticeable at all by a user

Reply:
Thanks for the detailed investigation! I'll go ahead and resolve this thread.

Just as a point of curiosity, I replicated this benchmark example using matrices instead of data frames. There was no meaningful or consistent difference between the different approaches, but thought I'd mention this because when matrices were involved instead of data frames, the run times were in units of seconds rather than microseconds. There must be some trick that R is using when it's working with data frames, where the data frame contents are left where they are in memory and list entries point to the same place? Whereas in contrast, creating a copy of a matrix (with or without shuffled columns) requires allocating new memory?? Not totally clear to me. But it does suggest that for matrices, there might be a way to optimize if no operation is required...

```
test_mat <- matrix(rnorm(3 * 10e7), ncol = 3)
colnames(test_mat) <- c("a", "b", "c")

lobstr::obj_size(test_mat)
#> 2.40 GB

order <- colnames(test_mat)
rev_order <- rev(colnames(test_mat))
random_order <- colnames(test_mat)[sample.int(ncol(test_mat), ncol(test_mat))]

microbenchmark::microbenchmark(
    correct_order =
        {test_mat[, order]},
    reverse_order = {
        test_mat[, rev_order]},
    random_order = {
        test_mat[, random_order]
    },
    check_first_correct_order = {
        if (isTRUE(all.equal(order, names(test_mat)))) {} else {test_mat[, order]}
    },
    check_first_reverse_order = {
        if (isTRUE(all.equal(rev_order, names(test_mat)))) {} else {test_mat[, rev_order]}
    },
    check_first_random_order = {
        if (isTRUE(all.equal(random_order, names(test_mat)))) {} else {test_mat[, random_order]}
    },
    ctrl = {
        test_mat[, ]
    }
)
```

Output:
```
Unit: seconds
                      expr      min       lq     mean   median       uq
             correct_order 2.211515 2.401191 2.454885 2.441378 2.517958
             reverse_order 2.242207 2.345728 2.436357 2.422561 2.510624
              random_order 2.202179 2.339214 2.445772 2.442194 2.511851
 check_first_correct_order 2.194511 2.361945 2.452138 2.438889 2.531446
 check_first_reverse_order 2.246732 2.362789 2.469375 2.453394 2.542674
  check_first_random_order 2.249296 2.364880 2.445531 2.429417 2.498123
                      ctrl 2.205778 2.381000 2.458361 2.449534 2.534418
      max neval
 2.883434   100
 2.772470   100
 2.860860   100
 2.850801   100
 3.003759   100
 3.201079   100
 2.790413   100
```